jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -1647,4 +1589,149 @@ private void complete()
+            flush();
+            return streamId;
+        }
+
+        private void newStream(HeadersFrame frame, Promise<Stream> promise, Stream.Listener listener)
+        {
+            Slot slot = new Slot();
+            try
+            {
+                int streamId = frame.getStreamId();
+                if (streamId <= 0)
+                {
+                    synchronized (this)
+                    {
+                        streamId = localStreamIds.getAndAdd(2);
+                        slots.offer(slot);
+                    }
+                    PriorityFrame priority = frame.getPriority();
+                    priority = priority == null ? null : new PriorityFrame(streamId, priority);
+                    frame = new HeadersFrame(streamId, priority, frame);
+                }
+                else
+                {
+                    synchronized (this)
+                    {
+                        slots.offer(slot);
+                    }
+                }
+
+                fillSlotWithNewLocalStream(slot, frame, promise, listener, streamId);
+            }
+            catch (Throwable x)
+            {
+                removeSlotAndFailPromise(slot, promise, x);
+            }
+        }
+
+        private void push(PushPromiseFrame frame, Promise<Stream> promise, Stream.Listener listener)
+        {
+            Slot slot = new Slot();
+            try
+            {
+                int streamId;
+                synchronized (this)
+                {
+                    streamId = localStreamIds.getAndAdd(2);
+                    slots.offer(slot);
+                }
+                frame = new PushPromiseFrame(streamId, frame);
+
+                fillSlotWithNewLocalStream(slot, frame, promise, listener, streamId);
+            }
+            catch (Throwable x)
+            {
+                removeSlotAndFailPromise(slot, promise, x);
+            }
+        }
+
+        private void fillSlotWithNewLocalStream(Slot slot, Frame frame, Promise<Stream> promise, Stream.Listener listener, int streamId)
+        {
+            IStream stream = createLocalStream(streamId);
+            stream.setListener(listener);
+            slot.entry = new ControlEntry(frame, stream, new StreamPromiseCallback(promise, stream));
+            flush();
+        }
+
+        private void removeSlotAndFailPromise(Slot slot, Promise<Stream> promise, Throwable x)
+        {
+            synchronized (this)
+            {
+                slots.remove(slot);
+            }
+            flush();
+            promise.failed(x);
+        }
+
+        private void flush()
+        {
+            Thread thread = Thread.currentThread();
+            boolean queued = false;
+            while (true)
+            {
+                ControlEntry entry = null;
+                synchronized (this)
+                {
+                    if (flushing != null && flushing != thread)
```

Comment:
I very much prefer the style of fields starting with `_` as it makes viewing code outside of an IDE so much better.   Jetty only stopped doing this after a hostile takeover of the style by @sbordet and @joakime :) 

Anyway, why do we need `flushing` to be a `Thread`?  Can't it just be a boolean as it is only changed inside the sync block?   In fact do we need it at all?  Any thread flushing will hold the lock until the queue is empty or it sees a null entry.   Any other thread calling `flush` may as well just poll rather than checking flushing?

Reply:
Even better: using an `AtomicBoolean` to track the flushing status, flush-skipping threads can bail out of that method without even acquiring the lock.

Reply:
I don't think it works. The exit condition must check the queue _and_ reset `flushing` atomically.

Reply:
@sbordet You're right; and since the check happens within the while loop, we must be able to discriminate between the first time the thread runs the loop and another thread noticing that the flushing in is progress. Hence, a ternary state is needed so the thread reference does that job.